source/server/connection_handler_impl.h

```diff
@@ -331,51 +334,100 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler,
   };
 
   using ActiveTcpListenerOptRef = absl::optional<std::reference_wrapper<ActiveTcpListener>>;
+  using UdpListenerCallbacksOptRef =
+      absl::optional<std::reference_wrapper<Network::UdpListenerCallbacks>>;
 
   struct ActiveListenerDetails {
     // Strong pointer to the listener, whether TCP, UDP, QUIC, etc.
     Network::ConnectionHandler::ActiveListenerPtr listener_;
-    // Reference to the listener IFF this is a TCP listener. Null otherwise.
-    ActiveTcpListenerOptRef tcp_listener_;
+
+    absl::variant<absl::monostate, std::reference_wrapper<ActiveTcpListener>,
+                  std::reference_wrapper<Network::UdpListenerCallbacks>>
+        typed_listener_;
```

Comment:
Just for my understanding, why do we need the variant? Is this just because it truly is a variant and it avoids multiple variables that are XOR? Maybe just add a small comment?

Reply:
It's just a safe `union`